src/Workspaces/Core/Portable/Classification/ClassificationServiceWithProviders.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Classification
+{
+    /// <summary>
+    /// A <see cref="ClassificationService"/> that defers to one or more <see cref="ClassificationProvider"/>.
+    /// </summary>
+    internal abstract class ClassificationServiceWithProviders : ClassificationService
+    {
+        private readonly Workspace _workspace;
+
+        protected ClassificationServiceWithProviders(Workspace workspace)
+        {
+            _workspace = workspace;
+        }
+
+        protected abstract string Language { get; }
+
+        private ImmutableArray<ClassificationProvider> _importedProviders;
+
+        protected ImmutableArray<ClassificationProvider> GetProviders()
+        {
+            if (_importedProviders == null)
+            {
+                var language = this.Language;
+                var mefExporter = (IMefHostExportProvider)_workspace.Services.HostServices;
+
+                var providers = ExtensionOrderer.Order(
+                        mefExporter.GetExports<ClassificationProvider, ClassificationProviderMetadata>()
+                        .Where(lz => lz.Metadata.Language == language)
+                        ).Select(lz => lz.Value).ToImmutableArray();
+
+                ImmutableInterlocked.InterlockedCompareExchange(ref _importedProviders, providers, default(ImmutableArray<ClassificationProvider>));
+            }
+
+            return _importedProviders;
+        }
+
+        public override ClassifiedSpan AdjustClassification(SourceText text, ClassifiedSpan classifiedSpan)
+        {
+            var extensionManager = _workspace.Services.GetService<IExtensionManager>();
+            foreach (var provider in this.GetProviders())
+            {
+                try
+                {
+                    if (!extensionManager.IsDisabled(provider))
+                    {
+                        var reclassified = provider.AdjustClassification(text, classifiedSpan);
+                        if (!reclassified.Equals(classifiedSpan))
+                        {
+                            return reclassified;
+                        }
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    throw;
+                }
+                catch (Exception e) when (extensionManager.CanHandleException(provider, e))
```

Comment:
We should be catching all provider exceptions using the extension manager for any providers that are implementable as a public API.

---
In reply to: [78663752](https://github.com/dotnet/roslyn/pull/13741#discussion_r78663752) [](ancestors = 78663752)